queries/java.scm

```diff
@@ -0,0 +1,161 @@
+    (try_with_resources_statement)
+    (while_statement)
+    (yield_statement)
+
+    ;; exceptions
+    ;; ";",
+    ;; "block",
+    (method_declaration)
+    (constructor_declaration)
+    (field_declaration)
+] @statement
+
+(class_declaration
+    name: (_) @className
+) @class @className.domain
+
+;;!! void myFunk() {}
+;;!  ^^^^^^^^^^^^^^^^
+(method_declaration
+    (identifier) @functionName
+) @namedFunction @functionName.domain
+(constructor_declaration
+    (identifier) @functionName
+) @namedFunction @functionName.domain
+
+;;!! ((value) -> true)
+;;!   ^^^^^^^^^^^^^^^
+(lambda_expression) @anonymousFunction
+
+;;!! if (value) {}
+;;!  ^^^^^^^^^^^^^
+(
+    (if_statement) @ifStatement
+    (#not-parent-type? @ifStatement "if_statement")
+)
+
+;;!! "string"
+;;!  ^^^^^^^^
+(string_literal) @string @textFragment
+
+;;!! // comment
+;;!  ^^^^^^^^^^
+[
+    (line_comment)
+    (block_comment)
+] @comment @textFragment
+
+;;!! int[] values = {1, 2, 3};
+;;!                 ^^^^^^^^^
+(array_initializer) @list
+
+;;!! List<String> value = new ArrayList() {{ add("a"); }};
+;;!                                       ^^^^^^^^^^^^^^^
+(object_creation_expression
+    (class_body
+        (block) @map
+    )
+)
+
+[
+    (method_invocation)
+    (object_creation_expression)
+    (explicit_constructor_invocation)
+] @functionCall
+
+;;!! case "0": return "zero";
+;;!  ^^^^^^^^^^^^^^^^^^^^^^^^
+;;!! case "0" -> "zero";
+;;!  ^^^^^^^^^^^^^^^^^^^
+[
+    (switch_block_statement_group)
+    (switch_rule)
+] @branch
+
+;;!! if () {}
+;;!  ^^^^^^^^
+(
+    (if_statement
+        "if" @branch.start @condition.domain.start
+        condition: (_) @condition
+        consequence: (block) @branch.end @condition.domain.end
+    ) @dummy
+    (#not-parent-type? @dummy "if_statement")
+    (#child-range! @condition 0 -1 true true)
+)
+
+;;!! else if () {}
+;;!  ^^^^^^^^^^^^^
+(if_statement
+    "else" @branch.start @condition.domain.start
+    alternative: (if_statement
+        condition: (_) @condition
+        consequence: (block) @branch.end @condition.domain.end
+        (#child-range! @condition 0 -1 true true)
+    )
+)
+
+;;!! else {}
+;;!  ^^^^^^^
+(if_statement
+    "else" @branch.start
+    alternative: (block) @branch.end
+)
+
+;;!! for (int i = 0; i < 5; ++i) {}
+;;!                  ^^^^^
+;;!  ------------------------------
+(for_statement
+    condition: (_) @condition
+) @_.domain
+
+;;!! while (value) {}
+;;!         ^^^^^
+;;!  ----------------
+(while_statement
+    condition: (_) @condition
+    (#child-range! @condition 0 -1 true true)
+) @_.domain
+
+;;!! switch (value) {}
+;;!          ^^^^^
+;;!  -----------------
+(switch_expression
+    condition: (_) @switchStatementSubject
+    (#child-range! @switchStatementSubject 0 -1 true true)
+) @_.domain
+
+;;!! true ? 1 : 2
+(ternary_expression
+    condition: (_) @condition
+    consequence: (_) @branch
+) @condition.domain
+(ternary_expression
+    alternative: (_) @branch
```

Comment:
Do we have a way to define a scope multiple places in the same pattern?

Reply:
unfortunately not. The problem is that then things like `.leading`, `.start`, etc become ambiguous. I guess we could allow it if we don't see any of those things in the pattern? 🤔 cc/ @josharian

Reply:
I have now forgotten all of the details but last time I tried playing games like that I ended up in a tangled mess of code. I will generally choose clear over clever/concise, personally.

Reply:
Couldn't we number the scopes? 
```
(ternary_expression
    consequence: (_) @branch1
    alternative: (_) @branch2
)
```

If we can't use actual numbers we could use letters or roman numerals? :D

Reply:
That's an interesting idea. Or `@branch.1`

Reply:
Yeah probably even better